README.md
# akronim

Library for doctest like functionality for Clojure programming language

[![CircleCI](https://circleci.com/gh/MichaelLeachim/akronim.svg?style=svg&circle-token=24d07c948681a700a9b64b6341de07357d0658b4)](https://circleci.com/gh/MichaelLeachim/akronim)


## Installation

[![Clojars Project](https://clojars.org/thereisnodot/akronim/latest-version.svg)](https://clojars.org/thereisnodot/akronim) 


## Usage

The akronim package provides `defns` macro 
to make testing and inplace documentation easier. 

### Getting started

```clojure
(require '[thereisnodot.akronim.core :refer [defns]))

;; Tests

(defns simple-sum
  "Will execute function over value of the map"
  [(simple-sum 3 4) => 7
   (simple-sum 4 5) => 9]
  [a b]
  (+ a b))
  
(test #'simple-sum) => :ok

;; On the other hand, erroneous tests will give AssertionError

(defns simple-sum-will-fail
  "Will execute function over value of the map"
  [(simple-sum 3 4) => 222]
  [a b]
  (+ a b))
  
(test #'simple-sum-will-fail) => 
AssertionError Assert failed: (clojure.core/= (test-defns-simple-sum-will-fail 3 4) 222) 

;; Examples. Providing examples in this form is useful for external documentation tools

(:akronim/example (meta #'simple-sum)) => 
  ["(simple-sum 3 4)" "=>" "7"] 
  ["(simple-sum 4 5)" "=>" "9"]
  
```


### Using with core.test

Here is how to use it in conjunction with testing pipeline. 

```clojure

(ns thereisnodot.utils.html-test
  (:require [clojure.test :refer :all]
            [thereisnodot.utils.html :as html]))
            
(deftest test-inlines-within-namespace
  (doseq [[symbol access] (ns-publics 'thereisnodot.utils.html)]
    (when (:test (meta access))
      (testing (str "Testing inlines of: " symbol)
        (is (=  (test access) :ok))))))
```

### Altering docstring

The `:akronim/example` metadata field is formatted with the help of  [zprint](https://github.com/kkinnear/zprint) library. 
By default, it is set to be **80** characters long. 

To alter docstring of a function at the time of declaration 
set the following config option either as: `.lein-env` or `.boot-env` or environment
variables or Java system properties. 



Set `akronim_docstring` parameter as:
* **nothing** will not alter docstring on declaration
* **text**    will append examples at the end of the docstring
* **markdown** will wrap the declaration into markdown Clojure code block and append examples  at the end of the docstring
* **hljs** will wrap the declaration into HTML code for https://highlightjs.org and append examples at the end of the docstring

Default is **nothing**


```clojure

(defns simple-sum
  "Will execute function over value of the map"
  [(simple-sum 3 4) => 7
   (simple-sum 4 5) => 9]
  [a b]
  (+ a b))
  
;; with nothing set
(:doc (meta #'simple-sum)) => "Will execute function over value of the map"

;; with text set
(:doc (meta #'simple-sum)) => 
"Will execute function over value of the map

(simple-sum 3 4) => 7
(simple-sum 4 5) => 9"

;; with markdown set
(:doc (meta #'simple-sum)) => 
"Will execute function over value of the map

```clojure
(simple-sum 3 4) => 7
(simple-sum 4 5) => 9
```"

;; with hljs set
(:doc (meta #'simple-sum)) => 
"Will execute function over value of the map</br>
<pre>
  <code class='cljs clojure'>
    (simple-sum 3 4) => 7
    (simple-sum 4 5) => 9
  </code>
</pre>
```

### :no-test?  metadata declaration

Sometimes we have to demonstrate function usage, but, for some
reason, we don't want the function to generate `:test` 

Usually, it happens when we are dealing with mutability. 

In this case, pass `:no-test?` as a metadata field to the 
function declaration. For example:

```clojure

(defns temp-file
  "Will return temp file handler without actually making it"
  {:no-test? true}
  [(str (temp-file)) => "/tmp/cljtmpfile1555885885588503551"]
  ([]
   (temp-file "cljtmpfile" ""))
  ([prefix extension]
   (java.io.File/createTempFile prefix extension)))

(test #'temp-file)
;; => :no-test
(:akronim/example (meta #'temp-file))
;; => (list [(str (temp-file)) => "/tmp/cljtmpfile1555885885588503551"])
```

## Development

This will clone the repo and start a local repl and run tests

```shell
git clone github.com/MichaelLeachim/akronim;
cd akronim;
tmuxinator . ;
```

## Change Log

### [Unreleased]

### [0.1.2]   2018-10-21
#### Added
- formatting for :akronim/example metadata field

#### Changed
- ~~There are tools that assume function declaration form to be `defn` not `defns`~~ 
  Now the same declaration as per `defn` should work. See tests
  

### [0.1.0] - 2018-10-20
#### Added
- configuration through environ library

#### Added
- basic functionality
- readme
- core tests
- template tests
- external configuration

#### Changed 
- Moved from another project into separate repository
- Moved templates

#### Removed 
- User defined :example and docstring generation funtionality

[0.1.2]: https://github.com/michaelleachim/akronim/compare/0.1.0...0.1.2
[Unreleased]: https://github.com/michaelleachim/akronim/compare/0.1.2..HEAD




CHANGELOG.md
# Change Log

## [Unreleased]

## [0.1.2]   2018-10-21
### Added
- formatting for :akronim/example metadata field

### Changed
- ~~There are tools that assume function declaration form to be `defn` not `defns`~~ 
  Now the same declaration as per `defn` should work. See tests
  

## [0.1.0] - 2018-10-20
### Added
- configuration through environ library



### Added
- basic functionality
- readme
- core tests
- template tests
- external configuration

### Changed 
- Moved from another project into separate repository
- Moved templates

### Removed 
- User defined :example and docstring generation funtionality

[0.1.2]: https://github.com/michaelleachim/akronim/compare/0.1.0...0.1.2
[Unreleased]: https://github.com/michaelleachim/akronim/compare/0.1.2..HEAD


test/thereisnodot/akronim/core_test.clj
;; @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
;; @ Copyright (c) Michael Leachim                                                      @
;; @ You can find additional information regarding licensing of this work in LICENSE.md @
;; @ You must not remove this notice, or any other, from this software.                 @
;; @ All rights reserved.                                                               @
;; @@@@@@ At 2018-17-10 23:32 <mklimoff222@gmail.com> @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

(ns ^{:doc "Testing for akronim package"
      :author "Michael Leahcim"}
  thereisnodot.akronim.core-test
  (:require [clojure.test :refer :all]
            [thereisnodot.akronim.test-helpers :as test-helpers]
            [thereisnodot.akronim.core :as doctest :refer
             [defns make-test-form make-example-form]]))

(deftest test-make-example-form
  (let [mult (fn [a b] (* a b))]
    (testing "Should work on let declared forms"
      (is
       (=
        (make-example-form
         (mult 3 2) => 6
         (mult 5 8) => 40)
        (list
         ["(mult 3 2)" "=>" "6"]
         ["(mult 5 8)" "=>" "40"]))))
    (testing "Formatting should work"
      (is
       (=
        (make-example-form
         {:hello {:my {:dear {:friend "whatever" :this "thing" :doesnt "work" "as" "expected"}}}} =>
         {:whatever [:whenever :what :else :should :work]})
        (list
       ["{:hello {:my {:dear {:doesnt \"work\",
                     :friend \"whatever\",
                     :this \"thing\",
                     \"as\" \"expected\"}}}}" "=>"
        "{:whatever [:whenever :what :else :should :work]}"]))))
    (testing "Should work on fns from other ns"
      (is
       (=
        (make-example-form
         (+ 3 4) => 7
         (* 2 9) => 18
         (- 3 3) => 0
         (/ 1 0) => 0)
        (list ["(+ 3 4)" "=>" "7" ]
              ["(* 2 9)" "=>" "18"]
              ["(- 3 3)" "=>" "0" ]
              ["(/ 1 0)" "=>" "0" ]))))))

(deftest test-make-test-form
  (let [mult (fn [a b] (* a b))]
    (testing "Basic workage"
      (is
       (=
        ((make-test-form
          (mult 3 2) => 6
          (mult 5 8) => 40)) nil)))
    (testing "Should raise an assert error"
      (is
       (thrown? AssertionError
                ((make-test-form
                  (mult 3 2) => 6
                  (mult 5 8) => 41)))))
    (testing "Should work on fns from other ns"
      (is
       (thrown? ArithmeticException
        ((make-test-form
          (+ 3 4) => 7
          (* 2 9) => 18
          (- 3 3) => 0
          (/ 1 0) => 0)))))))

(deftest test-defns
  (testing "Has example key"
    (is
     (=  (:akronim/example (meta #'test-helpers/test-defns-simple-sum))
         (list ["(test-defns-simple-sum 2 3)" "=>" "5"]
               ["(test-defns-simple-sum 3 3)" "=>" "6"]
               ["(test-defns-simple-sum 0 0)" "=>" "0"]))))
  (testing "should have examples, but not test. Because of `:no-test?` metadata key. But `:doc` shold be inplace"
    (is (=  (:akronim/example (meta #'test-helpers/no-test-fn))
            (list ["(no-test-fn 3 4)" "=>" "7"])))
    (is (=  (:doc (meta #'test-helpers/no-test-fn))
            "Should have an example, but no `:test` implementation"))
    (is (=  (:test (meta #'test-helpers/no-test-fn)) nil)))
  
  (testing "shouldn't have examples or test  because it is empty. "
    (is (=  (:akronim/example (meta #'test-helpers/should-have-no-examples)) nil))
    (is (=  (:test            (meta #'test-helpers/should-have-no-examples)) nil))
    (is (=  (:doc            (meta #'test-helpers/should-have-no-examples))
            "shouldn't have examples or test  because it is empty. ")))
 
  (testing "workage of different meta declarations"
    (let [meta-sym (meta #'test-helpers/different-meta-declarations)]
      (is (=  (:akronim/example meta-sym)
              (list ["(different-meta-declarations 3 3)" "=>" "9"])))
      (is (=  (:doc meta-sym) "Some text"))
      (is (=  (:echo meta-sym) true))
      (is (=  (:whatever meta-sym) "works"))))
  
  
  
  (testing "Test testing function"
    (is (thrown? AssertionError
                 (test #'test-helpers/test-defns-simple-mult-fail))
        "Should raise assertion error, as test is incorrect")
    (is (= (test #'test-helpers/test-defns-simple-sum) :ok) "Should pass"))
  
  (testing "Should also work from other fns"
    (is (=  (test  #'test-helpers/something-to-test) :ok))
    (is (thrown? AssertionError  (test  #'test-helpers/something-to-fail))))
  
  (testing "Should work on variadic functions"
    (is (thrown? AssertionError  (test #'test-helpers/variadic-fn)))))


src/thereisnodot/akronim/test_helpers.clj
;; @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
;; @ Copyright (c) Michael Leahcim                                                      @
;; @ You can find additional information regarding licensing of this work in LICENSE.md @
;; @ You must not remove this notice, or any other, from this software.                 @
;; @ All rights reserved.                                                               @
;; @@@@@@ At 2018-10-18 22:28 <thereisnodotcollective@gmail.com> @@@@@@@@@@@@@@@@@@@@@@@@
(ns
    ^{:doc "Inline asserts and helpers"
      :author "Michael Leahcim"}
    thereisnodot.akronim.test-helpers
  (:require [thereisnodot.akronim.core :refer [defns]]))

(defns no-test-fn
  "Should have an example, but no `:test` implementation"
  {:no-test? true}
  [(no-test-fn 3 4) => 7]
  ([a b]
   (+ a b)))

(defns should-have-no-examples
  "shouldn't have examples or test  because it is empty. "
  [] [] 3)

(defns
  ^{:whatever "works"}
  different-meta-declarations
  "Some text"
  {:echo true}
  [(different-meta-declarations 3 3) => 9]
  [a b]
  (* a b))

(defns variadic-fn
  "Takes several arguments as parameters"
  [(variadic-fn 1 1) => 2
   (variadic-fn 1 1 1) => 3
   (variadic-fn 1 1 1 1) => 5]
  ([a b]
   (variadic-fn a b 0 0))
  ([a b c]
   (variadic-fn a b c 0))  
  ([a b c d]
   (+ a b c d)))

(defns test-defns-simple-mult-fail
  "Multiplies two numbers. Tests should fail"
  [(test-defns-simple-mult-fail 2 3) => 6
   (test-defns-simple-mult-fail 3 3) => 9
   (test-defns-simple-mult-fail 0 1) => 12]
  [a b]
  (* a b))

(defns test-defns-simple-sum
  "Sums two numbers. Tests should pass"
  [(test-defns-simple-sum 2 3) => 5
   (test-defns-simple-sum 3 3) => 6
   (test-defns-simple-sum 0 0) => 0]
  [a b] (+ a b))

(defns ^{:what? "nothng"} test-defns-meta
  "Sums two numbers. Tests should pass"
  [(test-defns-simple-sum 2 3) => 5]
  [a b] (+ a b))

(defns something-to-test
  "Multtplies two numbers. Passes tests correctly"
  [(something-to-test 3 3) => 9
   (something-to-test 2 3) => 6
   (something-to-test 2 8) => 16]
  [a b]
  (* a b))

(defns something-to-fail
  "Sums two numbers. Fails tests"
  [(something-to-test 3 3) => 6
   (something-to-test 2 3) => 5
   (something-to-test 2 8) => 16]
  [a b]
  (+ a b))


test/thereisnodot/akronim/templates_test.clj
;; @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
;; @ Copyright (c) Michael Leahcim                                                      @
;; @ You can find additional information regarding licensing of this work in LICENSE.md @
;; @ You must not remove this notice, or any other, from this software.                 @
;; @ All rights reserved.                                                               @
;; @@@@@@ At 2018-10-18 23:38 <thereisnodotcollective@gmail.com> @@@@@@@@@@@@@@@@@@@@@@@@

(ns ^{:doc "Testing for akronim templates"
      :author "Michael Leahcim"}
  thereisnodot.akronim.templates-test
  (:require [clojure.test :refer :all]
            [thereisnodot.akronim.templates :as doctest-templates]))

;; TODO:
;; * set up templating through: https://github.com/weavejester/environ,
;;   :edn "akronim_markdown|akronim_html|akronim_base"
;; * fix up readme

(def case-list
  ["(hello 3 4)"  "=>" "7"
   "(hello 4 4)"  "=>" "8"
   "(hello 2 -2)" "=>" "0"])

(deftest test-docstring-template-nothing
  (testing "General usage"
    (is (=  (doctest-templates/docstring-template-nothing "Hello world" case-list)
            "Hello world"))))
(deftest test-docstring-template-text
  (testing "General usage"
    (is (=  (doctest-templates/docstring-template-text "Hello world" case-list)
            "Hello world\n\n(hello 3 4) => 7\n(hello 4 4) => 8\n(hello 2 -2) => 0\n"))))

(deftest test-docstring-template-markdown
  (testing "General usage"
    (is (=  (doctest-templates/docstring-template-markdown "Hello world" case-list)
            "Hello world\n\n```clojure\n\n(hello 3 4) => 7\n(hello 4 4) => 8\n(hello 2 -2) => 0\n\n```"))))

(deftest test-docstring-template-hljs
  (testing "General usage"
    (is (=   (doctest-templates/docstring-template-hljs "Hello world" case-list)
            "Hello world</br><pre><code class='hljs clojure'>\n(hello 3 4) => 7\n(hello 4 4) => 8\n(hello 2 -2) => 0\n</code></pre>"))))


src/thereisnodot/akronim/templates.clj
;; @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
;; @ Copyright (c) Michael Leahcim                                                      @
;; @ You can find additional information regarding licensing of this work in LICENSE.md @
;; @ You must not remove this notice, or any other, from this software.                 @
;; @ All rights reserved.                                                               @
;; @@@@@@ At 2018-10-18 23:40 <thereisnodotcollective@gmail.com> @@@@@@@@@@@@@@@@@@@@@@@@

(ns thereisnodot.akronim.templates)

(defn docstring-template-nothing
  [docstring examples] docstring)
(defn docstring-template-text
  [docstring examples]
  (str docstring "\n\n"
       (apply
        str
        (for [[should _ result] (partition 3 examples)]
          (str  should " => " result "\n")))))

(defn docstring-template-markdown
  [docstring examples]
  (str docstring "\n\n"
       "```clojure\n\n"
       (apply
        str
        (for [[should _ result] (partition 3 examples)]
          (str  should " => " result "\n")))
       "\n```"))

(defn docstring-template-hljs
  [docstring examples]
  (str docstring "</br>"
       "<pre><code class='hljs clojure'>\n"
       (apply
        str
        (for [[should _ result] (partition 3 examples)]
          (str  should " => " result "\n")))
       "</code></pre>"))


src/thereisnodot/akronim/core.clj
;; @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
;; @ Copyright (c) Michael Leahcim                                                      @
;; @ You can find additional information regarding licensing of this work in LICENSE.md @
;; @ You must not remove this notice, or any other, from this software.                 @
;; @ All rights reserved.                                                               @
;; @@@@@@ At 2018-10-18 19:47 <thereisnodotcollective@gmail.com> @@@@@@@@@@@@@@@@@@@@@@@@
(ns
    ^{:doc "Inline asserts and helpers"
      :author "Michael Leahcim"}
    thereisnodot.akronim.core
  (:require [thereisnodot.akronim.templates :as tmpl]
            [environ.core :as environ]
            [zprint.core :as zp]))

(def docstring-append
  (condp = (environ/env :akronim-docstring)
    "nothing"  tmpl/docstring-template-nothing
    "text"     tmpl/docstring-template-text
    "markdown" tmpl/docstring-template-markdown
    "hljs"     tmpl/docstring-template-hljs
    tmpl/docstring-template-nothing))

(defmacro make-test-form
  "Will turn user friendly description into an assert/test form"
  [& input]
  `(fn []
    ~(cons
      `do
      (for [[k _ v] (partition 3 input)]
        `(assert (=  ~k ~v))))))

(defmacro make-example-form
  "Will turn input forms into strings for documentation"
  [& input]
  (cons
   list
   (for [[k _ v] (partition 3 input)]
     [(zp/zprint-str k 80) "=>" (zp/zprint-str v 80)])))

;; (defmacro defns
;;   "will add examples in the form of :example and :test to the
;;    metadata"
;;   [fname docstring examples & decls]
;;   (list* `defn (with-meta fname
;;                  (assoc (meta fname)
;;                         :akronim/example
;;                         `(make-example-form ~@examples)
;;                         :test
;;                         (if (:no-test? (meta fname)) nil `(make-test-form ~@examples))
;;                         :doc             (docstring-append docstring examples)))
;;          decls))

(defmacro defns
  "will add examples in the form of :example and :test to the
   metadata"
  [fname & decls]
  (let [;; if string, then docstring, so add
        m (if (string? (first decls)) {:doc (first decls)} {})
        ;; if string then docstring, so skip
        decls  (if (string? (first decls)) (next decls) decls)
        ;; if map, then metadata, so add
        m (if (map? (first decls)) (conj m (first decls)) m)
        ;; if map, then metada, so skip
        decls  (if (map? (first decls)) (next decls) decls)
        ;; if vec, then examples, so add
        examples (if (vector? (first decls)) (first decls) [])
        ;; if vector, then examples, so skip
        decls (if (vector? (first decls)) (next decls) decls)
        m (merge (meta fname) m)
        ;; rest goes to original defn function
        ]
    (if-not (empty? examples)
      (list* `defn (with-meta fname
                     (assoc m
                            :akronim/example
                            `(make-example-form ~@examples)
                            :test
                            (if (:no-test? m) nil `(make-test-form ~@examples))
                            :doc             (docstring-append (:doc m) examples)))
             decls)
      ;; when no examples, let the function be the usual
      (list* `defn (with-meta fname m) decls))))

(defmacro defns-
  "same as defn, yielding non-public def"
  [name & decls]
  (list* `defns (with-meta name (assoc (meta name) :private true)) decls))
